lib/src/set/set_builder.dart

```diff
@@ -170,12 +208,15 @@ class SetBuilder<E> {
 
   Set<E> get _safeSet {
     if (_setOwner != null) {
-      _set = new Set<E>.from(_set);
+      _set = _createSet()..addAll(_set);
       _setOwner = null;
     }
     return _set;
   }
 
+  Set<E> _createSet() =>
+      _setConstructor != null ? _setConstructor() : new Set<E>();
```

Comment:
I think it's probably worth getting the generic type right.

I had to play around with dartpad to check that this works, but you can ask for a

```
Set<T> Function<T>()
```

the only downside is that people can't declare one inline--at least, I don't know a syntax that lets a lambda take a type parameter. So they'll have to declare their set factories at top level, so:

https://dartpad.dartlang.org/02c460fdd95b918527d12ebd57044f9e

Note that in dart 1 the type will not be used, i.e. the sets will always be `<dynamic>`, but in Dart 2/DDC, the type _is_ used and correct at runtime.

What do you think?

Reply:
Sorry, I don't understand what you mean. Are you talking about `_setConstructor`/`_setFactory`, or about the `new Set<E>()` expression? They both have a generic type, do they not?

If you pass a function of the wrong generic type to `withBase()`, it will be reported as a static error: https://dartpad.dartlang.org/ebbda240657f2075bedbfdbae06c12c6

Reply:
I think you need

```
_setConstructor<E>()
```

here, so `_setConstructor` can create a Set<E> instead of a Set<dynamic>. To do that, the type of `_setConstructor` will need to be `Set<T> Function<T>()` instead of `Set<T> Function()`.

Reply:
My thought process was: At the moment I instantiate a builder, I *must* know its generic type, so I already know the set type that `base` must return, and therefore there is no need to keep the factory generic.

I assumed `base` would look like this:
```
Set<String> stringSetFactory() => new SplayTreeSet<String>();
final intSetFactory = () => new SplayTreeSet<int>();
```
These functions obviously return a typed set, not a `dynamic` one. Then you'd use them like this:
```
final ascendingNumbers = new BuiltSet<int>.build((b) => b
  ..withBase(intSetFactory)
  ..addAll([0, 3, 2, 5]));
```
If you try `withBase(stringSetFactory)`, you'll see the static analysis warning _The argument type '() → Set\<String\>' can't be assigned to the parameter type '() → Set\<int\>'_, because for this exact builder, `E` is statically known to be `int`.

Or am I missing something here?

Reply:
Ah, I missed that. Yes, you're correct. Thanks!